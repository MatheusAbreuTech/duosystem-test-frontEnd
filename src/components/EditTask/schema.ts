import { z } from 'zod'

export const schema = z.object({
  editTaskField: z.string().min(1, 'Informe uma descrição para a tarefa.')
})
