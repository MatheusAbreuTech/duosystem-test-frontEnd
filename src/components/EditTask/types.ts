import { z } from 'zod'
import { schema } from './schema'

export type EditTaskProps = {
  taskId: string
  setEditTaskId: (taskId: string) => void
}

export type SchemaType = z.infer<typeof schema>
