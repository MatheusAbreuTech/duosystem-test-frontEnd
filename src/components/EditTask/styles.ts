import theme from '@/styles/theme'
import styled from 'styled-components'

export const Form = styled.form`
  display: flex;
  flex-direction: column;
  align-items: flex-start;
`

export const Wrapper = styled.div`
  display: flex;
  align-items: center;
  gap: 1rem;

  > input {
    border: 1px solid ${theme.colors.black};
  }
`
