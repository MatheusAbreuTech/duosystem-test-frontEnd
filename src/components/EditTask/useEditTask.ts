import useTaskStore from '@/store/zustand'
import { SubmitHandler, useForm } from 'react-hook-form'
import { EditTaskProps, SchemaType } from './types'
import { schema } from './schema'
import { zodResolver } from '@hookform/resolvers/zod'

export const useEditTask = ({ taskId, setEditTaskId }: EditTaskProps) => {
  const { taskList, changeTaskDescriptionById } = useTaskStore()

  const methods = useForm<SchemaType>({
    resolver: zodResolver(schema),
    defaultValues: {
      editTaskField: taskList.find((task) => task.id === taskId)?.description
    }
  })

  const {
    handleSubmit,
    reset,
    formState: { errors }
  } = methods

  const handleEditTask: SubmitHandler<SchemaType> = (data) => {
    const { editTaskField } = data

    changeTaskDescriptionById(taskId, editTaskField)
    setEditTaskId('')
    reset()
  }
  return {
    errors,
    methods,
    handleSubmit,
    handleEditTask
  }
}
