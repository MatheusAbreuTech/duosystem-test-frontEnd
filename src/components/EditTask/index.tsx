'use client'

import * as S from './styles'
import Input from '../Input'
import Button from '../Button'
import { FormProvider } from 'react-hook-form'

import ErrorMessage from '../ErrorMessage'
import { useEditTask } from './useEditTask'
import { EditTaskProps } from './types'

const EditTask = ({ taskId, setEditTaskId }: EditTaskProps) => {
  const { methods, errors, handleSubmit, handleEditTask } = useEditTask({
    taskId,
    setEditTaskId
  })

  return (
    <FormProvider {...methods}>
      <S.Form
        onSubmit={handleSubmit(handleEditTask)}
        data-testId={`editTaskComponent_${taskId}`}
      >
        <S.Wrapper>
          <Input name="editTaskField" data-testId={`editTaskField_${taskId}`} />
          <Button
            type="submit"
            color="green"
            label="Salvar"
            data-testId={`editTaskButton_${taskId}`}
          />
        </S.Wrapper>
        <ErrorMessage
          errors={errors}
          field="editTaskField"
          data-testId={`editTaskError_${taskId}`}
        />
      </S.Form>
    </FormProvider>
  )
}

export default EditTask
