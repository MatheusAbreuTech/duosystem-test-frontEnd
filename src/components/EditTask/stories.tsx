import { Meta, StoryObj } from '@storybook/react'
import EditTask from '.'

export default {
  title: 'EditTask',
  component: EditTask
} as Meta

export const Default: StoryObj = {}
