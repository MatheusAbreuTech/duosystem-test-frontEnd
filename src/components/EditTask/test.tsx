import { render } from '@testing-library/react'

import EditTask from '.'

describe('<EditTask />', () => {
  it('should render the EditTask component', () => {
    const setEditTaskIdMock = jest.fn()
    const { container } = render(
      <EditTask taskId={1} setEditTaskId={setEditTaskIdMock} />
    )

    expect(container).toBeInTheDocument()

    expect(container.firstChild).toMatchSnapshot()
  })
})
