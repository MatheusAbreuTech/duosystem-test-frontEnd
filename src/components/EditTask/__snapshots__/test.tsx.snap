// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EditTask /> should render the EditTask component 1`] = `
.c0 {
  display: flex;
  flex-direction: column;
  align-items: flex-start;
}

.c1 {
  display: flex;
  align-items: center;
  gap: 1rem;
}

.c1 >input {
  border: 1px solid #383838;
}

.c2 {
  width: 100%;
  height: 42px;
  padding: 0 1rem;
  background-color: #fff;
  outline: none;
  border: 1px solid #383838;
  font-size: 1.6rem;
  border-radius: 5px;
}

.c3 {
  background-color: #3AC922;
  color: #ffffff;
  border: none;
  border-radius: 0.5rem;
  padding: 1rem;
  max-height: 4.2rem;
  height: 100%;
  cursor: pointer;
  font-size: 1.6rem;
}

<form
  action=""
  class="c0"
  data-testid="editTaskComponent_id"
>
  <div
    class="c1"
  >
    <input
      class="c2"
      data-testid="editTaskField_id"
      name="editTaskField"
    />
    <button
      class="c3"
      color="green"
      data-testid="editTaskButton_id"
      type="submit"
    >
      Salvar
    </button>
  </div>
</form>
`;
